orangecontrib/prototypes/widgets/owimagegrid.py

```diff
@@ -0,0 +1,1141 @@
+                else:
+                    url = self.urlFromValue(inst[attr])
+                    thumbnail.setToolTip(url.toString())
+
+                    if url.isValid() and url.isLocalFile():
+                        reader = QImageReader(url.toLocalFile())
+                        image = reader.read()
+                        if image.isNull():
+                            error = reader.errorString()
+                            thumbnail.setToolTip(
+                                thumbnail.toolTip() + "\n" + error)
+
+                            self._errcount += 1
+                        else:
+                            pixmap = QPixmap.fromImage(image)
+                            thumbnail.setPixmap(pixmap)
+                            self._successcount += 1
+
+                        future = Future()
+                        future.set_result(image)
+                        future._reply = None
+                    elif url.isValid():
+                        future = self.loader.get(url)
+
+                        @future.add_done_callback
+                        def set_pixmap(future, thumb=thumbnail):
+                            if future.cancelled():
+                                return
+
+                            assert future.done()
+
+                            if future.exception():
+                                # Should be some generic error image.
+                                pixmap = QPixmap()
+                                thumb.setToolTip(thumb.toolTip() + "\n" +
+                                                 str(future.exception()))
+                            else:
+                                pixmap = QPixmap.fromImage(future.result())
+
+                            thumb.setPixmap(pixmap)
+
+                            self._noteCompleted(future)
+                    else:
+                        future = None
+
+                self.items.append(_ImageItem(i, thumbnail, url, future))
+
+            if any(not it.future.done() if it.future else False for it in self.items):
+                self.info.setText("Retrieving...\n")
+            else:
+                self._updateStatus()
+
+    def urlFromValue(self, value):
+        variable = value.variable
+        origin = variable.attributes.get("origin", "")
+        if origin and QDir(origin).exists():
+            origin = QUrl.fromLocalFile(origin)
+        elif origin:
+            origin = QUrl(origin)
+            if not origin.scheme():
+                origin.setScheme("file")
+        else:
+            origin = QUrl("")
+        base = origin.path()
+        if base.strip() and not base.endswith("/"):
+            origin.setPath(base + "/")
+
+        if os.path.exists(str(value)):
+            url = QUrl.fromLocalFile(str(value))
+        else:
+            name = QUrl(str(value))
+            url = origin.resolved(name)
+        if not url.scheme():
+            url.setScheme("file")
+        return url
```

Comment:
This looks messy. I posit the following snippet works the same for all inputs.
```py
def valueToQUrl(self, value):
    base = value.variable.attributes.get('origin', '')
    if QDir(base).exists():
        base = QUrl.fromLocalFile(base)
    else:
        base = QUrl(base)
    url = base.resolved(QUrl(value))
    return url
```
Don't we have this somewhere already?

Reply:
Using this instead of the old function, thanks.